src/alert_manager.cpp

```diff
@@ -152,54 +175,147 @@ namespace libtorrent
 	void alert_manager::set_notify_function(boost::function<void()> const& fun)
 	{
 		mutex::scoped_lock lock(m_mutex);
-		m_notify = fun;
-		if (!m_alerts[m_generation].empty())
+
+		// in order to make boost:function<> atomic we use two separate
+		// fields in a ping-pong fashion and update the pointer to it
+		// atomically
+		if (fun)
+		{
+			const int index = (m_notify == &m_notify_storage[0]) ? 1 : 0;
+			m_notify_storage[index] = fun;
+			m_notify.store(&m_notify_storage[index]);
+		}
+		else
+		{
+			m_notify.store(NULL);
```

Comment:
Yes, but it's tricky because if you do it too soon while another thread still has it loaded there's a problem. This is true even if the load is not relaxed on the other thread because there's still be a period between loading and consuming it.

I think I tried to make it a shared_ptr and had some issues with it. But you right something should be done.

Reply:
Actually boost::function<void()> should only hold pointers to the function and to the object (the this pointer) right? AFAICT it is the caller's responsibility to ensure that the object is not destroyed before the session or before another callback is set.

But there is a problem here though. If the caller destroys the object immediately after calling this function it is possible that `emplace_alert()` still has a reference to the old one (probably that's what you meant). The solution is to wait until `m_buffer_refs == 0` like we do for `m_dispatch`. We also need to move the `m_buffer_refs` increment to the top of `emplace_alert()` to solve the same problem with `set_dispatch_function()`.

Reply:
I was referring to the closure that may be captured by the function object. The client may have members that it expects will be destructed in a timely fashion. For example, it may have a `shared_ptr<>` that keeps the `alert_manager` object alive (creating a cycle unless it gets destructed when cleared).

Reply:
Ok, got it. On the current version it is fixed by using mutual exclusion, by using a shared mutex. We take an exclusive lock before the functions or resizing the queue, so it will make sure that all threads are out of emplace_alert. I think it's much easier to read now and less complex. All functions work just like before except that many threads can execute `emplace_alert()` concurrently and `pop_alerts()` does not block threads emplacing alerts.